### PR TITLE
The final copy refactor

### PR DESCRIPTION
**Pull request by dgaubert**
The main objective is to split the code in 2 parts:
- Expressjs and streams
- Database / pgstream

### CODE REVIEW

**Comment by oleurud** [comment]
Ready to CR @dgaubert @enekid @rafatower

File: app/controllers/copy_controller.js [hunk 1]
```diff
@@ -58,59 +61,91 @@ CopyController.prototype.route = function (app) {
...
         res.header("Content-Type", "application/octet-stream");
 
-        streamCopy.to(
-            res, 
-            req.query.q, 
-            res.locals.userDbParams,
-            res.locals.user,
-            logger,
-            function(err) {
+        streamCopy.getPGStream(StreamCopy.ACTION_TO, function (err, pgstream) {
```
Comment (enekid) [inline app/controllers/copy_controller.js]: why not an arrow function?

```javascript
streamCopy.getPGStream(StreamCopy.ACTION_TO, (err, pgstream) => {
```

File: app/controllers/copy_controller.js [hunk 2]
```diff
@@ -58,59 +61,91 @@ CopyController.prototype.route = function (app) {
...
                 if (err) {
                     return next(err);
                 }
-                
-                // this is a especial endpoint
-                // the data from postgres is streamed to response directly
+
+                pgstream
+                    .on('data', data => metrics.addSize(data.length))
+                    .on('error', (err) => {
```
Comment (enekid) [inline app/controllers/copy_controller.js]: `()` not needed

File: app/controllers/copy_controller.js [hunk 3]
```diff
@@ -58,59 +61,91 @@ CopyController.prototype.route = function (app) {
...
-            logger,
-            function(err, metrics) {
+        const sql = req.query.q;
+        const { userDbParams, user } = res.locals;
+        const isGzip = req.get('content-encoding') === 'gzip';
+
+        const streamCopy = new StreamCopy(sql, userDbParams);
+        const metrics = new StreamCopyMetrics(logger, 'copyfrom', sql, user, isGzip);
+
+        streamCopy.getPGStream(StreamCopy.ACTION_FROM, function (err, pgstream) {
```
Comment (enekid) [inline app/controllers/copy_controller.js]: again `err => {}`?

**Pull request merged**